server/raft.go

```diff
@@ -1160,13 +1161,31 @@ func (n *raft) isCatchingUp() bool {
 	return n.catchup != nil
 }
 
-// Lock should be held.
+// Lock should be held. This function may block for up to ~5ms to check
+// forward progress in some cases.
 func (n *raft) isCurrent() bool {
-	// First check if have had activity and we match commit and applied.
-	if n.commit == 0 || n.commit != n.applied {
-		n.debug("Not current, commit %d != applied %d", n.commit, n.applied)
-		return false
+	// First check if have had activity and that commit and applied are equal.
+	// If not, wait for a short period of time and see if we are making forward
+	// progress.
+	startDelta := n.commit - n.applied
+	if n.commit == 0 || startDelta > 0 {
+		var progress bool
+		for i := 0; i < 10; i++ { // 5ms, in 0.5ms increments
+			n.RUnlock()
+			time.Sleep(time.Millisecond / 2)
+			n.RLock()
+			if curDelta := n.commit - n.applied; curDelta < startDelta {
+				// The gap is getting smaller, so we're making forward progress.
+				progress = true
+				break
```

Comment:
Could just return true here, and if we fail out just warn after loop. No need for progress bool now.

Reply:
Is it OK to skip the later checks by doing that, i.e. that either we're the leader and/or have heard from the leader recently?

Reply:
Right, so the way the loop is set up at the moment is if there's no `progress` then we return false, otherwise we just fall through to the leader checks.

Reply:
We should do leader checks first and only check deltas if we are a follower and we know we have been in recent contact with the leader.

My bad there, did not catch that.

Reply:
OK, I've refactored it to do leader/follower checks first.